### PR TITLE
Xray 

### LINKED ISSUE #289: Xray
Reopening, Xray got removed in #1510

Maybe with an opacity option as well, suggested in #964

### PR DESCRIPTION
**Pull request by S-B99**
Closes #289 
Unblocks #11, #394 and #429

### CODE REVIEW

File: src/main/java/me/zeroeightsix/kami/command/commands/XRayCommand.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    public void call(String[] args) {
+        XRay xr = (XRay) ModuleManager.getModuleByName("XRay");
+        if (xr == null) {
+            Command.sendChatMessage("&cThe XRay module is not available for some reason.");
+            return;
+        }
+        if (!xr.isEnabled()) {
+            Command.sendChatMessage("&cWarning: The XRay module is not enabled.");
+            Command.sendChatMessage("&cThese commands will still have effects, but will not visibly do anything.");
+        }
```
Comment (S-B99) [inline src/main/java/me/zeroeightsix/kami/command/commands/XRayCommand.java]: Reminder for myself to change this to send warning message

File: src/main/java/me/zeroeightsix/kami/command/commands/XRayCommand.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        setDescription("Has a set of sub-commands to control the XRay module.");
+    }
+
+    @Override
+    public void call(String[] args) {
+        XRay xr = (XRay) ModuleManager.getModuleByName("XRay");
+        if (xr == null) {
+            Command.sendChatMessage("&cThe XRay module is not available for some reason.");
+            return;
+        }
```
Comment (S-B99) [inline src/main/java/me/zeroeightsix/kami/command/commands/XRayCommand.java]: This to send error message

File: src/main/java/me/zeroeightsix/kami/command/commands/XRayCommand.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import net.minecraft.block.Block;
+
+/**
+ * Created by 20kdc on 17/02/2020.
+ */
+public class XRayCommand extends Command {
+    public XRayCommand() {
+        super("xray", new ChunkBuilder().append("subcommand").build());
+        setDescription("Has a set of sub-commands to control the XRay module.");
+    }
```
Comment (S-B99) [inline src/main/java/me/zeroeightsix/kami/command/commands/XRayCommand.java]: Better description and aliases

**Comment by S-B99** [comment]
Also add a xray thing to .help if it doesn't already have a proper help I'm sleepy

**Comment by S-B99** [comment]
Also, if a `minecraft:` isn't specified autocomplete it when saving and removing from the list

**Comment by 20kdc** [comment]
The `minecraft:` prefix is optional due to the way the get-block-by-name function works; it's sanitized on addition or removal of elements from the list using the `.xray +block -block` syntax.

**Comment by S-B99** [comment]
Oh it already is?

**Comment by 20kdc** [comment]
Yes, but since it's an internal detail of the setting the relevant code is in the XRay module itself.

**Comment by S-B99** [comment]
Oooh ok, thank you, that's good to know.

**Comment by S-B99** [comment]
Now this is only waiting for 1 merge conflict and to change the stuff to sendErrorMessage and everything. I'll go eat something first I guess

**Comment by 20kdc** [comment]
To be specific:
https://github.com/S-B99/kamiblue/pull/447/files/9c774ffc93cbea58e5e363401ea4c2dd5f090c42#diff-7c8abe7a172750487fdbbfb9dffc8557R73

1. If a value is added to the list, it may end up without the `minecraft:` prefix, but it gets sanitized on the next pass, and the relevant functions handle it there.
2. If a value is removed from the list, it's removed by-Block-instance so how it's expressed or any aliases don't matter.
3. When the list is being read, it's (again) converted to Block instances.

**Comment by 20kdc** [comment]
A note: The name of the block states/etc. is dependent on mod ID, which might cause trouble for the merge.

**Comment by S-B99** [comment]
Huh, where is that defined? And why is it dependant on it?
Unless you modified the modid lines it should be fine?

**Comment by 20kdc** [comment]
The XRay defines a block. Check resources. The Java code will be fine, but the block JSON has `kami:` references all over it...

**Comment by 20kdc** [comment]
Er, just to be clear, the current state of feature/master will probably malfunction as the `kami` references have not been changed to `kamiblue`.
These are the `resources/assets/kami` directory and the JSON files in there.

**Comment by S-B99** [comment]
If you see in #289 I just changed the references, but I should be using kamiblue for the folder name as well. Just curious, why did you do those json things instead of using a ResourceLocation?

**Comment by 20kdc** [comment]
If ResourceLocations are able to dynamically define blocks without any JSON support files, then it'd be good to switch to that; I wasn't aware of it and went with the "default" method for adding a new block.

**Comment by S-B99** [comment]
Ah. I'm not sure if how exactly this works, I haven't looked to closely at it but basically you can just open a resource location to a file and use it as a texture, look at how InventoryPreview does it.

**Comment by S-B99** [comment]
https://github.com/S-B99/kamiblue/blob/feature/master/src/main/java/me/zeroeightsix/kami/module/modules/gui/InventoryViewer.java#L134

**Comment by 20kdc** [comment]
Using it as a texture, yes, but the XRay doesn't actually do any rendering, as that could cause problems with OptiFine/etc.
Instead the XRay changes which block states the renderer gets. This implies anything that the XRay sends to the renderer has to be in the form of a valid block state.

**Comment by S-B99** [comment]
oooh ok. makes sense now c:

**Comment by nehemiasjoelwww** [comment]
yesss

**Pull request merged**